### ISSUE TITLE
Group Matching Backend

### ISSUE THREAD

**Issue by tamaraniac** [rjaltman/dogu#40]
We have decided that one of our advanced functionalities will be to match students into project groups, based on their interest in certain project ideas. There's still a lot to think about here:

- If too many students put a project as their first choice, how is it decided who gets that project?
- How do we make sure every group is of sufficient size? (if only one student preferences a project, even if that project is their first choice, it should be skipped over, etc)
- Do we ever match a student with a course they didn't rank? (This would likely be necessary in any class) How is that decided?
- When do we drop a consistently ranked project, because a different grouping of students would simply be more ideal numbers wise?

Basically, this will need some sort of algorithm, far beyond just SQL queries. It's basically the [stable marriage](https://en.wikipedia.org/wiki/Stable_marriage_problem) problem, but with an uneven mapping. Maybe closer to the [assignment problem](https://en.wikipedia.org/wiki/Assignment_problem). Either way, this will be a lot of optimization. I'm going to do some thinking.
